### PR TITLE
Add k8s master feature

### PR DESCRIPTION
**Pull request by lixiaoyuner**
Signed-off-by: Yun Li <yunli1@microsoft.com>

#### Why I did it
Add some packages which k8s master image requires. 
#### How I did it
Install some packages
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

### CODE REVIEW

File: build_debian.sh [hunk 1]
```diff
@@ -253,24 +253,50 @@ sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install docker-ce=${DOCKER_VERSIO
...
-    ## Install Kubernetes
-    echo '[INFO] Install kubernetes'
+install_kubernetes () {
     sudo https_proxy=$https_proxy LANG=C chroot $FILESYSTEM_ROOT curl -fsSL \
         https://packages.cloud.google.com/apt/doc/apt-key.gpg | \
         sudo LANG=C chroot $FILESYSTEM_ROOT apt-key add -
     ## Check out the sources list update matches current Debian version
     sudo cp files/image_config/kubernetes/kubernetes.list $FILESYSTEM_ROOT/etc/apt/sources.list.d/
     sudo LANG=C chroot $FILESYSTEM_ROOT apt-get update
+    sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install kubelet=${1}-00
```
Comment (qiluo-msft) [inline build_debian.sh:263]: >${1} [](http://example.com/codeflow?start=67&length=4)

Define a named variable at function beginning to help reader.

```
local ver="$1"
```

File: files/build_templates/sonic_debian_extension.j2 [hunk 2]
```diff
@@ -731,6 +731,27 @@ sudo https_proxy=$https_proxy LANG=C chroot $FILESYSTEM_ROOT docker $SONIC_NATIV
 echo "docker images pull complete"
 {% endif %}
 
+{% if include_kubernetes_master == "y" %}
+# Pull in kubernetes master docker images
+echo "pulling kubernetes master images ..."
+sudo https_proxy=$https_proxy LANG=C chroot $FILESYSTEM_ROOT docker $SONIC_NATIVE_DOCKERD_FOR_DOCKERFS pull k8s.gcr.io/pause:${MASTER_PAUSE_VERSION}
+sudo https_proxy=$https_proxy LANG=C chroot $FILESYSTEM_ROOT docker $SONIC_NATIVE_DOCKERD_FOR_DOCKERFS pull k8s.gcr.io/kube-apiserver:v${KUBERNETES_VERSION}
```
Comment (qiluo-msft) [inline files/build_templates/sonic_debian_extension.j2:738]: >KUBERNETES_VERSION [](http://example.com/codeflow?start=137&length=18)

Is it MASTER_KUBERNETES_VERSION?

File: build_debian.sh (hunk 1 above)
Comment (lixiaoyuner) [reply build_debian.sh]: Have done

File: files/build_templates/sonic_debian_extension.j2 (hunk 2 above)
Comment (lixiaoyuner) [reply files/build_templates/sonic_debian_extension.j2]: Yes, it should be MASTER_KUBERNETES_VERSION, I copied the wrong line. Have corrected it to MASTER_KUBERNETES_VERSION.

File: Makefile.work [hunk 3]
```diff
@@ -366,6 +367,11 @@ SONIC_BUILD_INSTRUCTION :=  make \
                            KUBERNETES_VERSION=$(KUBERNETES_VERSION) \
                            KUBERNETES_CNI_VERSION=$(KUBERNETES_CNI_VERSION) \
                            K8s_GCR_IO_PAUSE_VERSION=$(K8s_GCR_IO_PAUSE_VERSION) \
+                           INCLUDE_KUBERNETES_MASTER=$(INCLUDE_KUBERNETES_MASTER) \
+                           MASTER_KUBERNETES_VERSION=$(MASTER_KUBERNETES_VERSION) \
+                           MASTER_PAUSE_VERSION=$(MASTER_PAUSE_VERSION) \
+                           MASTER_COREDNS_VERSION=$(MASTER_COREDNS_VERSION) \
+                           MASTER_ETCD_VERSION=$(MASTER_ETCD_VERSION) \
```
Comment (xumia) [inline Makefile.work:374]: We should not pass the K8s detail package version info by the sonic slave command line.
You can add a config file in rules, for instance, rules/k8s.mk, the make file will be loaded into slave.mk automatically.

File: build_debian.sh [hunk 4]
```diff
@@ -253,24 +253,55 @@ sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install docker-ce=${DOCKER_VERSIO
...
-    echo '[INFO] Install kubernetes'
+install_kubernetes () {
+    local ver="$1"
     sudo https_proxy=$https_proxy LANG=C chroot $FILESYSTEM_ROOT curl -fsSL \
         https://packages.cloud.google.com/apt/doc/apt-key.gpg | \
         sudo LANG=C chroot $FILESYSTEM_ROOT apt-key add -
     ## Check out the sources list update matches current Debian version
     sudo cp files/image_config/kubernetes/kubernetes.list $FILESYSTEM_ROOT/etc/apt/sources.list.d/
     sudo LANG=C chroot $FILESYSTEM_ROOT apt-get update
+    sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install kubelet=${ver}-00
```
Comment (xumia) [inline build_debian.sh:264]: The suffix -00 is part of the version number, how about changing the version.
If the kubectl version change to 1.21.1-01, the script will be broken.
```
KUBERNETES_VERSION = 1.21.1
```
To
```
KUBERNETES_VERSION = 1.21.1-00
```

**Review by xumia — APPROVED** [review]

File: rules/config [hunk 5]
```diff
@@ -174,10 +174,25 @@ KUBE_DOCKER_PROXY = http://172.16.1.1:3128/
 # These are Used *only* when INCLUDE_KUBERNETES=y
 # NOTE: As a worker node it has to run version compatible to kubernetes master.
 #
-KUBERNETES_VERSION = 1.21.1
-KUBERNETES_CNI_VERSION = 0.8.7
+KUBERNETES_VERSION = 1.21.1-00
+KUBERNETES_CNI_VERSION = 0.8.7-00
 K8s_GCR_IO_PAUSE_VERSION = 3.4.1
```
Comment (qiluo-msft) [inline rules/config:179]: Seem you do not need to move this block.

Comment (lixiaoyuner) [reply rules/config:179]: Xuhui has a comment, he said the version number is not correct, adding the postfix should be good, he asked me help change this code.

File: rules/config [hunk 6]
```diff
@@ -174,10 +174,25 @@ KUBE_DOCKER_PROXY = http://172.16.1.1:3128/
 # These are Used *only* when INCLUDE_KUBERNETES=y
 # NOTE: As a worker node it has to run version compatible to kubernetes master.
 #
-KUBERNETES_VERSION = 1.21.1
-KUBERNETES_CNI_VERSION = 0.8.7
+KUBERNETES_VERSION = 1.21.1-00
```
Comment (renukamanavalan) [inline rules/config:177]: Can we upgrade client version to latest too?

**Review by renukamanavalan — APPROVED** [review]

File: rules/config (hunk 5 above)
Comment (lixiaoyuner) [reply rules/config]: Changed

File: build_debian.sh (hunk 4 above)
Comment (lixiaoyuner) [reply build_debian.sh]: Changed

File: Makefile.work (hunk 3 above)
Comment (lixiaoyuner) [reply Makefile.work]: Changed

File: rules/config (hunk 6 above)
Comment (lixiaoyuner) [reply rules/config]: Upgraded

**Review by qiluo-msft — APPROVED** [review]

**Comment by qiluo-msft** [comment]
This commit could not be cleanly cherry-picked to 202012. Please submit another PR.

**Comment by lixiaoyuner** [comment]
Add wrong Approved label, remove it and add Request label.

**Pull request merged**